### PR TITLE
Fix empty relationships

### PR DESCRIPTION
**Pull request by dmarkow**
Here's my first attempt at fixing #26, though it's definitely not ready to merge yet.

I agree on the integration test, so I started with that. I added `has_many :likes` and `has_one :excerpt` and set them both to be empty. As expected, the test fails.

To handle a blank `has_one`, I updated the `ResourceIdentifier` formatter to format an empty relationship as `nil` (instead of a resource identifier object with a blank `id` field, e.g. ```{"type": "excerpt", "id"; ""}```).

Next, I updated `put_if_present` in the `Utils` module to explicitly add `nil`/ `[]` to `dict` only when it's for the `:data` key. However:

1. While the JSON API spec only requires explicit `null`/`[]` values when dealing with the `data` object, I don't really like explicitly matching on the `:data` key. But maybe I'm just overthinking it.
2. More importantly, because of this approach (specifically, the guard clause on the `nil` case), `{"data": ...}` will be added to **every** relationship in the output, even if we didn't ask it to (e.g. we may have just asked for `has_one :excerpt, link: "..."`). And even worse, if we didn't ask for `:data`, it'll be set to `null` even if there really are items in the relationship.

So now I'm stuck. The `JaSerializer.Builder.Relationship` struct includes the `:data` key, which defaults to `nil`. If we have an empty has_one relationship, that also sets `:data` to nil. So how do we tell the difference between "it's nil because we never set it to anything" vs "it's nil because that's specifically what we want"? (This is why I originally suggested using a `:null_value` atom, though I agree that doesn't feel right).

### CODE REVIEW

File: lib/ja_serializer/formatter/resource_identifier.ex [hunk 1]
```diff
@@ -1,4 +1,5 @@
 defimpl JaSerializer.Formatter, for: JaSerializer.Builder.ResourceIdentifier do
+  def format(%{id: nil}), do: nil
```
Comment (alanpeabody) [inline lib/ja_serializer/formatter/resource_identifier.ex]: :+1:

**Comment by alanpeabody** [comment]
Thanks! I think this looks good. I will pull it down today/tonight, see if something strikes me that would help with the issues you raised (particularly the 2nd one), and get back to you with feedback.

**Comment by alanpeabody** [comment]
Okay, so after tracing through this I have actually come back to either using your option special atom option or extending the struct to contain more context (eg `%Relationship{includes_data: true}`).

I actually think with a couple tweaks the atom option is pretty good:

My initial reaction was that `:null_value` felt wrong because it seems so close to the existing `:nil` atom. I think however if we change the name to be more representative it "feels" better. Something like `:not_found` and maybe even `[:not_found]`.

I think this could be added in easily to the case statement in JaSerializer.Builder.ResourceIdentifier you identified in #26 .

```elixir
[] -> [:not_found]
nil -> :not_found
```

The other thing I would probably consider is not using Utils.put_if_present for the `:data` in the Relationship formatter. I think a custom `add_data_if_present` or something similar that isolates the `:not_found` handling logic would help reveal the intent of the code.

I just played briefly with this, so I may have overlooked something.

Thanks for digging into this and contributing!

**Comment by dmarkow** [comment]
Nice, that worked out perfectly:

- Added `:not_found` and `[:not_found]` to the builder (Note: I might be more inclined to use the atom `:empty_relationship` since that's inline with how JSON API refers to it?)
- Got rid of the `def format(%{id: nil}), do: nil` line I originally added in the `ResourceIdentifier` formatter since we don't need it anymore (there will no longer be a situation where `id` is `nil`)
- Added a new `add_data_if_present` that explicitly adds `nil`/`[]` for `:not_found`/`[:not_found]`, otherwise it defers to `put_if_present`
- Also updated the tests to make sure we're not including `data: null` in the formatted output when we never asked for `:data` in the first place.

I think this now meets all of the requirements for dealing with empty relationships. If everything looks good to you, I can squash these commits so it can be merged. (If you prefer pull requests include updates to the changelog, I can include that too). Thanks!!!

**Comment by alanpeabody** [comment]
This is great, thank you!

I can update the change log when I push out a new release.

If you want to change the atom to `:empty_relationship` I agree that is better then not found.

I will merge this up today either way and push out a release this weekend (I want to hit the other bug you addressed and spend a little more focused time with `mix profile.fprof` to make sure there are no terrible bottlenecks in the code).

**Comment by dmarkow** [comment]
Updated and squashed. I really appreciate your time helping me get this working. I'm very new to Elixir (coming from a long-standing Ruby mindset) and I feel like working through this helped me learn a bunch about pattern matching, protocols, etc.

**Comment by alanpeabody** [comment]
Awesome, I have also learned a lot building this library.  This has been my first use of protocols, and I may have admittedly gone a little over the top with macros to begin with but it has been really interesting.

I would be interesting in your feedback on the code in general, as well as how it is working for you as a library.

**Comment by dmarkow** [comment]
Haven't had much of a chance to use it so far -- it's my first time using the JSON API format, so I'm not even at the point that I have a client ready to consume my API, and I ran into the empty relationship issue early on. But the serializer API is very familiar coming from ActiveModelSerializers in Ruby, so it was easy to dive right in. And having absurdly fast response times helps too.

I'll probably have some more time next week to work on my project, so I'll have more feedback soon (and depending on how complex my API gets, I may be interested in the sparse fieldset/relationship stuff and might be able to help get that going)
